lambda_builders/actions/python_pip/DESIGN.md

```diff
@@ -0,0 +1,149 @@
+## Python - PIP Lambda Builder
+
+### Scope
+
+This package is an effort to port the Chalice packager to a library that can
+be used to handle the dependency resilution portion of packaging Python code
+for use in AWS Lambda. The scope for this builder is to take an existing
+directory containing customer code, and a top-levle `requirements.txt` file
+specifying third party depedencies. The builder will examine the dependencies
+and use pip to build and include the dependencies in the customer code bundle
+in a way that makes them importable.
+
+### Challenges
+
+Python is particularly difficult to package for other platforms because it is
+heavily coupled to your local environment. This is because `setup.py` the
+"config" to describe an install is just a python file, which means authors can
+run code to check things about your current system, and then use that to make
+decisions about how to install the package. These decisions are obviously not
+valid once we move the built package to a different platform.
+
+Python packaging also has the concept of a wheel, which is more of a self
+contained package that can be trivially installed into a python environment
+and does not execute any code. Our goal is to build up a set of these that
+are known to be compatible with AWS Lambda.
+
+### Interface
+
+The top level interface will consist of a single function with the signature:
+
+```python
+def build_depenencies(source_path, config=None):
+    """Builds a source directory's dependencies into the directory.
+
+	:type source_path: str
+	:param source_path: The source directory that needs to have dependencies
+		built for it. It is assumed that this customer code is correct and
+		would function if it had all of its dependencies installed alongside
+		it. It is assumed that the dependencies are specified in a top level
+		``requirements.txt`` file. This file can be removed before deploying
+		to lambda, but the this function will not modify any existing files
+		in this directory.
+
+	:type config: :class:`lambda_builders.actions.python_pip.config`
+	:param config: A configuration object which contains additional config
```

Comment:
I would call this out as optional, and not implemented, for now. This requirement exists for several build actions. So I want to standardize this as Optimizations that can turned on/off by customers - https://github.com/awslabs/aws-lambda-builders/issues/3

Reply:
Ah yes I meant to write that, my bad.

Reply:
Ah I missed it ;)